## **EDA de TODOS los años: (2015 - 2019)**

### En esta parte vamos a ver la comparación de los datasets de cada año para ver si existen pequeñas diferencias entre ellos y para revisar que transformaciones podemos hacer en el siguiente notebook

In [2]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns

### Leer los datos **(CSV)**

In [3]:
data_path = '../data/'

csv_files = {year: os.path.join(data_path, f'{year}.csv') for year in ['2015', '2016', '2017', '2018', '2019']}

dfs = {year: pd.read_csv(filepath) for year, filepath in csv_files.items()}

# 2015

In [ ]:
df_2015 = dfs['2015']
df_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 12 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        158 non-null    object 
 1   Region                         158 non-null    object 
 2   Happiness Rank                 158 non-null    int64  
 3   Happiness Score                158 non-null    float64
 4   Standard Error                 158 non-null    float64
 5   Economy (GDP per Capita)       158 non-null    float64
 6   Family                         158 non-null    float64
 7   Health (Life Expectancy)       158 non-null    float64
 8   Freedom                        158 non-null    float64
 9   Trust (Government Corruption)  158 non-null    float64
 10  Generosity                     158 non-null    float64
 11  Dystopia Residual              158 non-null    float64
dtypes: float64(9), int64(1), object(2)
memory usage: 1

In [5]:
df_2015.head()

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

In [6]:
df_2015.isnull().sum()

Country                          0
Region                           0
Happiness Rank                   0
Happiness Score                  0
Standard Error                   0
Economy (GDP per Capita)         0
Family                           0
Health (Life Expectancy)         0
Freedom                          0
Trust (Government Corruption)    0
Generosity                       0
Dystopia Residual                0
dtype: int64

No existen valores nulos

In [7]:
df_2015.duplicated().sum()

np.int64(0)

# 2016

In [8]:
df_2016 = dfs['2016']
df_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        157 non-null    object 
 1   Region                         157 non-null    object 
 2   Happiness Rank                 157 non-null    int64  
 3   Happiness Score                157 non-null    float64
 4   Lower Confidence Interval      157 non-null    float64
 5   Upper Confidence Interval      157 non-null    float64
 6   Economy (GDP per Capita)       157 non-null    float64
 7   Family                         157 non-null    float64
 8   Health (Life Expectancy)       157 non-null    float64
 9   Freedom                        157 non-null    float64
 10  Trust (Government Corruption)  157 non-null    float64
 11  Generosity                     157 non-null    float64
 12  Dystopia Residual              157 non-null    flo

In [9]:
df_2016.head()

Country          Region  Happiness Rank  Happiness Score  \
0      Denmark  Western Europe               1            7.526   
1  Switzerland  Western Europe               2            7.509   
2      Iceland  Western Europe               3            7.501   
3       Norway  Western Europe               4            7.498   
4      Finland  Western Europe               5            7.413   

   Lower Confidence Interval  Upper Confidence Interval  \
0                      7.460                      7.592   
1                      7.428                      7.590   
2                      7.333                      7.669   
3                      7.421                      7.575   
4                      7.351                      7.475   

   Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                   1.44178  1.16374                   0.79504  0.57941   
1                   1.52733  1.14524                   0.86303  0.58557   
2                   1.42666  1.18326                   0.86733  0.56624   
3                   1.57744  1.12690                   0.79579  0.59609   
4                   1.40598  1.13464                   0.81091  0.57104   

   Trust (Government Corruption)  Generosity  Dystopia Residual  
0                        0.44453     0.36171            2.73939  
1                        0.41203     0.28083            2.69463  
2                        0.14975     0.47678            2.83137  
3                        0.35776     0.37895            2.66465  
4                        0.41004     0.25492            2.82596

In [10]:
df_2016.isnull().sum()

Country                          0
Region                           0
Happiness Rank                   0
Happiness Score                  0
Lower Confidence Interval        0
Upper Confidence Interval        0
Economy (GDP per Capita)         0
Family                           0
Health (Life Expectancy)         0
Freedom                          0
Trust (Government Corruption)    0
Generosity                       0
Dystopia Residual                0
dtype: int64

No existen valores nulos

In [11]:
df_2016.duplicated().sum()

np.int64(0)

# 2017

In [12]:
df_2017 = dfs['2017']
df_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 12 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        155 non-null    object 
 1   Happiness.Rank                 155 non-null    int64  
 2   Happiness.Score                155 non-null    float64
 3   Whisker.high                   155 non-null    float64
 4   Whisker.low                    155 non-null    float64
 5   Economy..GDP.per.Capita.       155 non-null    float64
 6   Family                         155 non-null    float64
 7   Health..Life.Expectancy.       155 non-null    float64
 8   Freedom                        155 non-null    float64
 9   Generosity                     155 non-null    float64
 10  Trust..Government.Corruption.  155 non-null    float64
 11  Dystopia.Residual              155 non-null    float64
dtypes: float64(10), int64(1), object(1)
memory usage: 

In [13]:
df_2017.head()

Country  Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
0       Norway               1            7.537      7.594445     7.479556   
1      Denmark               2            7.522      7.581728     7.462272   
2      Iceland               3            7.504      7.622030     7.385970   
3  Switzerland               4            7.494      7.561772     7.426227   
4      Finland               5            7.469      7.527542     7.410458   

   Economy..GDP.per.Capita.    Family  Health..Life.Expectancy.   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   
3                  1.564980  1.516912                  0.858131  0.620071   
4                  1.443572  1.540247                  0.809158  0.617951   

   Generosity  Trust..Government.Corruption.  Dystopia.Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715  
3    0.290549                       0.367007           2.276716  
4    0.245483                       0.382612           2.430182

In [14]:
df_2017.isnull().sum()

Country                          0
Happiness.Rank                   0
Happiness.Score                  0
Whisker.high                     0
Whisker.low                      0
Economy..GDP.per.Capita.         0
Family                           0
Health..Life.Expectancy.         0
Freedom                          0
Generosity                       0
Trust..Government.Corruption.    0
Dystopia.Residual                0
dtype: int64

No existen valores nulos

In [15]:
df_2017.duplicated().sum()

np.int64(0)

# 2018

In [16]:
df_2018 = dfs['2018']
df_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     155 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB


In [17]:
df_2018.head()

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.632           1.305           1.592   
1             2            Norway  7.594           1.456           1.582   
2             3           Denmark  7.555           1.351           1.590   
3             4           Iceland  7.495           1.343           1.644   
4             5       Switzerland  7.487           1.420           1.549   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   
3                    0.914                         0.677       0.353   
4                    0.927                         0.660       0.256   

   Perceptions of corruption  
0                      0.393  
1                      0.340  
2                      0.408  
3                      0.138  
4                      0.357

In [18]:
df_2018.isnull().sum()

Overall rank                    0
Country or region               0
Score                           0
GDP per capita                  0
Social support                  0
Healthy life expectancy         0
Freedom to make life choices    0
Generosity                      0
Perceptions of corruption       1
dtype: int64

Hay un valor nulo en Perceptions of corruption, luego veremos que podemos hacer con el

In [19]:
df_2018.duplicated().sum()

np.int64(0)

# 2019

In [20]:
df_2019 = dfs['2019']
df_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     156 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB


In [21]:
df_2019.head()

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   
3             4           Iceland  7.494           1.380           1.624   
4             5       Netherlands  7.488           1.396           1.522   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   
3                    1.026                         0.591       0.354   
4                    0.999                         0.557       0.322   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                      0.341  
3                      0.118  
4                      0.298

In [22]:
df_2019.isnull().sum()

Overall rank                    0
Country or region               0
Score                           0
GDP per capita                  0
Social support                  0
Healthy life expectancy         0
Freedom to make life choices    0
Generosity                      0
Perceptions of corruption       0
dtype: int64

No existen valores nulos

In [23]:
df_2019.duplicated().sum()

np.int64(0)

### Comparación de las columnas en cada dataframe

In [ ]:
rows = []
for year, df in dfs.items():
    rows.append({
        "Año": year,
        "Columnas": "\n".join(df.columns.tolist())
    })

tabla = pd.DataFrame(rows, columns=["Año", "Columnas"])


print(tabla.to_string(index=False))

 Año                                                                                                                                                                                                                                  Columnas
2015                                       Country\nRegion\nHappiness Rank\nHappiness Score\nStandard Error\nEconomy (GDP per Capita)\nFamily\nHealth (Life Expectancy)\nFreedom\nTrust (Government Corruption)\nGenerosity\nDystopia Residual
2016 Country\nRegion\nHappiness Rank\nHappiness Score\nLower Confidence Interval\nUpper Confidence Interval\nEconomy (GDP per Capita)\nFamily\nHealth (Life Expectancy)\nFreedom\nTrust (Government Corruption)\nGenerosity\nDystopia Residual
2017                                    Country\nHappiness.Rank\nHappiness.Score\nWhisker.high\nWhisker.low\nEconomy..GDP.per.Capita.\nFamily\nHealth..Life.Expectancy.\nFreedom\nGenerosity\nTrust..Government.Corruption.\nDystopia.Residual
2018                                        

# **Resumen**

### 📈 Comparación Resumida de Datasets de Felicidad Mundial

Los cinco datasets analizados sobre el ranking mundial de felicidad comparten una estructura muy similar, pero presentan diferencias importantes en cantidad de datos, nombres de columnas y presencia de valores nulos. A continuación, se resume su comparación.

---

### 🔁 Similitudes

- Todos incluyen **métricas** como:
  - Puntaje de felicidad (`Happiness Score` o `Score`)
  - PIB per cápita (`GDP per Capita`)
  - Libertad (`Freedom`)
  - Generosidad (`Generosity`)
  - Apoyo social o familiar (`Family` o `Social support`)
  - Percepción de corrupción (`Trust` o `Perceptions of corruption`)
- La estructura general se mantiene:
  - Un identificador por país o región
  - Un ranking numérico de felicidad
  - La mayoría de los datos están completos (sin valores nulos)
  - No existen valores duplicados

---

### 🔀 Diferencias

#### 📌 Número de filas
- **Dataset 2015**: 158  
- **Dataset 2016**: 157
- **Dataset 2017**: 155
- **Datasets 2018 y 2019**: 156 
- **Total**: 784

#### 📊 Número de columnas
- **Dataset 2015**: 12  
- **Dataset 2016**: 13  
- **Dataset 2017**: 12  
- **Datasets 2018 y 2019**: 9

#### 🏷️ Nombres de columnas
- **Datasets 2015 y 2016**: Nombres descriptivos y detallados (e.g., `Economy (GDP per Capita)`)
- **Dataset 2017**: Nombres con notación por puntos (e.g., `Economy..GDP.per.Capita.`)
- **Datasets 2018 y 2019**: Nombres más cortos y limpios (e.g., `GDP per capita`, `Social support`)

#### ➕ Métricas adicionales
- **Dataset  2015**: Incluye `Standard Error`
- **Dataset 2016**: Incluye `Lower` y `Upper Confidence Interval`
- **Dataset 2017**: Incluye `Whisker.high` y `Whisker.low`
- **Datasets 2018 y 2019**: Solo incluyen variables principales; no incluyen métricas de intervalo ni error

---

### ⚖️ Variación entre Datasets

#### Mayor variación:
- **Cantidad y tipo de columnas**: Datasets 2018 y 2019 tienen menos variables; Dataset 2016 incluye más métricas estadísticas.
- **Presencia de métricas específicas**:
  - Intervalos de confianza → Solo en Datasets 2016 y 2017  
  - Error estándar → Solo en Dataset 2015

#### Menor variación:
- **Variables principales** están presentes en todos los datasets (con nombres distintos)
- **Tipos de datos**:
  - `float64` → métricas numéricas (felicidad, PIB, salud, etc.)
  - `int64` → ranking
  - `object` → nombre del país o región

---

### ❗ Valores Nulos

- **Solo el Dataset del 2018** presenta **1 valor nulo** en la columna `Perceptions of corruption`
- **Los demás datasets** están completos


Veremos una última cosa, y es la parte de la columna de **Dystopia residual**

+ Se encuentra en los CSVs de 2015, 2016 y 2017, y no se encuentra en los otros 2 restantes.

+ No sabemos que es pero mirando los csv individualmente podemos deducir que parece ser la resta de score - la suma de las variables:

```
GDP per capita
Family or Social Support
Health (Life Expectancy)
Freedom
Trust (Goverment Corruption)
Generosity
```

Teniendo en cuenta eso, vamos a realizar un código para verificar si esto es correcto.

De ser correcto lo que haremos será ignorar por completo la columna de **Dystopia Residual** debido a que es un valor que depende del Score, y se supone que lo que queremos predecir es el Score, y si tenemos un valor que depende de manera directa del score lo que haremos será generarle un problema al modelo.



In [ ]:
common_renames = {
    'Happiness Score': 'Score',
    'Happiness.Score': 'Score',
    'Economy (GDP per Capita)': 'Economy',
    'Economy..GDP.per.Capita.': 'Economy',
    'GDP per capita': 'Economy',
    'Family': 'Support',
    'Social support': 'Support',
    'Health (Life Expectancy)': 'Health',
    'Health..Life.Expectancy.': 'Health',
    'Healthy life expectancy': 'Health',
    'Freedom': 'Freedom',
    'Freedom to make life choices': 'Freedom',
    'Trust (Government Corruption)': 'Trust',
    'Trust..Government.Corruption.': 'Trust',
    'Perceptions of corruption': 'Trust',
    'Generosity': 'Generosity',
    'Dystopia Residual': 'Dystopia Residual',
    'Dystopia.Residual': 'Dystopia Residual'
}

factors = ['Economy','Support','Health','Freedom','Generosity','Trust']

def load_and_clean(year, data_path='../data/'):
    filepath = os.path.join(data_path, f'{year}.csv')
    df = pd.read_csv(filepath)
    df = df.rename(columns=common_renames).copy()
    df['Year'] = int(year)
    return df

for year in ['2015','2016','2017']:
    df_year = load_and_clean(year)
    df_year['check'] = df_year[factors].sum(axis=1) + df_year['Dystopia Residual']
    df_year['diff'] = df_year['Score'] - df_year['check']
    print(f"{year} → Máxima desviación absoluta: {round(df_year['diff'].abs().max(), 6)}")

    df_year['sum_factors'] = df_year[factors].sum(axis=1) + df_year['Dystopia Residual']
    print(f"\n=== Primeras filas año {year} ===")
    print(df_year[['Country', 'Score', 'sum_factors']].head(5).to_string(index=False))
    print("\n" + "-"*50 + "\n")


2015 → Máxima desviación absoluta: 0.00051

=== Primeras filas año 2015 ===
    Country  Score  sum_factors
Switzerland  7.587      7.58696
    Iceland  7.561      7.56092
    Denmark  7.527      7.52708
     Norway  7.522      7.52222
     Canada  7.427      7.42694

--------------------------------------------------

2016 → Máxima desviación absoluta: 0.0005

=== Primeras filas año 2016 ===
    Country  Score  sum_factors
    Denmark  7.526      7.52560
Switzerland  7.509      7.50866
    Iceland  7.501      7.50139
     Norway  7.498      7.49758
    Finland  7.413      7.41349

--------------------------------------------------

2017 → Máxima desviación absoluta: 0.000499

=== Primeras filas año 2017 ===
    Country  Score  sum_factors
     Norway  7.537     7.537079
    Denmark  7.522     7.521835
    Iceland  7.504     7.503704
Switzerland  7.494     7.494366
    Finland  7.469     7.469203

--------------------------------------------------



Aquí hacemos una última validación que confirma la hipótesis de que Dystopia es la variable Calculada como: 

Score - La suma de las otras 6 variables.

Por ese motivo su correlación es muy fuerte, pues es una variable que depende del valor de las otras variables que también explican Score.

+ **Es una variable dependiente de la columna objetivo que queremos calcular, por ese motivo debe ser eliminada**

Considerando todo lo anterior ahora lo que haremos será en un **próximo notebook** realizar las estandarizaciones necesarias para manejar nombres de columnas, trabajar los datos nulos, buscar la manera de integrar la columna **Region** que no está presente en todos los datasets. Entre otras cosas que veremos en el notebook de **002_transform_merge_eda.ipynb**

Luego de realizar las **estandarizaciones**, podriamos ver las correlaciones entre las variables.

Lo haremos con ese enfoque de unificación, porque me parece que de cierta manera **estandarizar y visualizar los datos de manera general**, es mucho más cómodo que analizar dataset por dataset cuando se trata de correlaciones.